File: deploy/cloudflare/workers/shared/fallback.ts
export type FallbackOutcome = "recovered" | "degraded" | "exhausted";

export type WorkerFallbackEvent = {
  component: "rate_limit" | "stt" | "other";
  from:
    | "durable_object"
    | "d1"
    | "fcm"
    | "workers_ai_streaming"
    | "workers_ai_classifier"
    | "restrict"
    | "none";
  to:
    | "durable_object_retry"
    | "deepgram_cloud"
    | "conservative_no_action"
    | "notification_outbox"
    | "throttle"
    | "unlimited"
    | "none";
  reason:
    | "dependency_unavailable"
    | "invalid_response"
    | "malformed_doc"
    | "malformed_response"
    | "other";
  outcome: FallbackOutcome;
  requestId?: string;
};

export function recordFallback(event: WorkerFallbackEvent): void {
  console.warn(
    JSON.stringify({
      event: "fallback",
      component: event.component,
      from: event.from,
      to: event.to,
      reason: event.reason,
      outcome: event.outcome,
      ...(event.requestId ? { request_id: event.requestId } : {}),
    }),
  );
}

File: deploy/cloudflare/python/api-ai/src/fallback.py
"""Bounded structured fallback telemetry for the Python Worker."""

from __future__ import annotations

import json


def record_fallback(*, from_mode: str, to_mode: str, reason: str, outcome: str) -> None:
    allowed_from = {"d1", "restrict", "none"}
    allowed_to = {"throttle", "none"}
    allowed_reasons = {"dependency_unavailable", "malformed_doc", "other"}
    allowed_outcomes = {"recovered", "degraded", "exhausted"}
    print(
        json.dumps(
            {
                "event": "fallback",
                "component": "other",
                "from": from_mode if from_mode in allowed_from else "none",
                "to": to_mode if to_mode in allowed_to else "none",
                "reason": reason if reason in allowed_reasons else "other",
                "outcome": outcome if outcome in allowed_outcomes else "degraded",
            },
            separators=(",", ":"),
            sort_keys=True,
        )
    )
